upgrading/topics/keycloak/changes-21_0_0.adoc

```diff
@@ -0,0 +1,75 @@
+= Context and dependency injection no longer enabled to JAX-RS Resources
+
+In order to provide a better runtime and leverage as much as possible the underlying stack,
+all injection points for contextual data using the `javax.ws.rs.core.Context` annotation were removed. The expected improvement
+in performance involves no longer creating proxies instances multiple times during the request lifecycle, and drastically reducing the amount of reflection calls at runtime.
+
+This change is also about changing our underlying JAX-RS implementation from Resteasy Classic to Resteasy Reactive.
+
+If you are extending one of the following SPIs:
+
+* `PolicySpi`
+* `AdminRealmResourceSpi`
+* `IdentityProviderSpi`
+* `RealmResourceProvider`
+
+You should review your custom JAX-RS (sub)resources in order obtain any contextual data from a `org.keycloak.common.util.ResteasyProvider`
+instance as follows:
+
+[source,java]
+----
+@Context
+org.jboss.resteasy.spi.HttpRequest request;
+
+@Context
+org.jboss.resteasy.spi.HttpResponse response;
+
+----
+was replaced by
++
+[source,java]
+----
+KeycloakSession session = // obtain the session, usually available when creating a custom provider from a factory
+KeycloakContext context = session.getContext();
+
+HttpRequest request = context.getHttpRequest();
+HttpResponse response = context.getHttpResponse();
+----
++
+
+In case you don't have access to a `KeycloakSession` instance when invoking a JAX-RS resource method, it is also possible to obtain
+contextual data from the JAX-RS runtime as follows:
+
++
+[source,java]
+----
+KeycloakSession session = Jaxrs.getContextData(KeycloakSession.class);
+----
++
+
+You are still able to resolve contextual data from the JAX-RS runtime by:
+
++
+[source,java]
+----
+KeycloakSession session = // obtain the session
+KeycloakContext context = session.getContex();
+MyContextualObject myContextualObject = context.getContextObject(MyContextualObject.class);
+----
++
+
+= Upgrading your custom JAX-RS resources
+
+If you are extending the server's REST APIs through the following SPIs:
+
+* `AdminRealmResourceSpi`
+* `IdentityProviderSpi`
+* `RealmResourceProvider`
+* `RealmResourceProvider`
+
+You need to add a `META-INF/beans.xml` to the JAR file where your custom providers are packaged. Otherwise, they won't be recognized by the server
```

Comment:
```suggestion
You need to add a `META-INF/beans.xml` to the JAR file where your custom providers are packaged. Otherwise, they are not recognized by the server
```